main.cpp

```diff
@@ -0,0 +1,60 @@
+#include <iostream>
+#include <thread>
+#include <vector>
+#include <future>
+
+std::mutex mtx;
+std::condition_variable cv;
+bool finished = false;
+
+class Producer {
+private:
+    std::thread::id thread_id;
+    double value;
+
+public:
+    Producer(std::thread::id id, double val) : thread_id(id), value(val) {}
+
+    void operator()() {
+        std::unique_lock<std::mutex> lock(mtx);
```

Comment:
Locking the mutex as soon as the producer starts means that in practice every producer will wait for the others. Your producers won't run in parallel and thus defeat the entire point of the threads.